tests/End2EndTestCase.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/**
+ * This file is part of the contentful.php package.
+ *
+ * @copyright 2015-2017 Contentful GmbH
+ * @license   MIT
+ */
+
+namespace Contentful\Tests\Delivery;
+
+use Contentful\Delivery\Client;
+use Contentful\Log\ArrayLogger;
+use Symfony\Component\Filesystem\Filesystem;
+
+class End2EndTestCase extends \PHPUnit_Framework_TestCase
+{
+    const CACHE_DIR = __DIR__ . '/../build/cache';
+
+    /**
+     * @var Client[]
+     */
+    private static $clients = [];
+
+    public static function setUpBeforeClass()
+    {
+        self::$clients = [
```

Comment:
What's the reason for the eager creation? Intuitively I'd have implemented this the same as the CMA, it's easier to reason about.